modules/saml/src/Auth/Process/AttributeNameID.php

```diff
@@ -60,35 +73,43 @@ public function __construct(array $config, $reserved)
      */
     protected function getValue(array &$state): ?string
     {
-        if (
-            !isset($state['Attributes'][$this->identifyingAttribute])
-            || count($state['Attributes'][$this->identifyingAttribute]) === 0
-        ) {
-            Logger::warning(
-                'Missing attribute ' . var_export($this->identifyingAttribute, true) .
-                ' on user - not generating attribute NameID.',
-            );
-            return null;
-        }
-        if (count($state['Attributes'][$this->identifyingAttribute]) > 1) {
-            Logger::warning(
-                'More than one value in attribute ' . var_export($this->identifyingAttribute, true) .
-                ' on user - not generating attribute NameID.',
-            );
-            return null;
+        foreach ($this->identifyingAttributes as $attr) {
+            if (
+                !isset($state['Attributes'][$attr])
+                || count($state['Attributes'][$attr]) === 0
+            ) {
+                Logger::warning(
+                    'Missing attribute ' . var_export($attr, true) .
+                    ' on user - not using for attribute NameID.',
+                );
+                continue;
+            }
+            if (count($state['Attributes'][$attr]) > 1) {
+                Logger::warning(
+                    'More than one value in attribute ' . var_export($attr, true) .
+                    ' on user - not using for attribute NameID.',
+                );
+                continue;
+            }
+            // just in case the first index is no longer 0
+            $value = array_values($state['Attributes'][$attr]);
+            $value = strval($value[0]);
+
+            if (empty($value)) {
+                Logger::warning(
+                    'Empty value in attribute ' . var_export($attr, true) .
+                    ' on user - not using for attribute NameID.',
+                );
+                continue;
+            }
+            // If we're still here, we found an attribute value for NameID
+            break;
         }
-        // just in case the first index is no longer 0
-        $value = array_values($state['Attributes'][$this->identifyingAttribute]);
-        $value = strval($value[0]);
+        unset($attr);
 
-        if (empty($value)) {
-            Logger::warning(
-                'Empty value in attribute ' . var_export($this->identifyingAttribute, true) .
-                ' on user - not generating attribute NameID.',
-            );
+        if (!isset($value)) {
```

Comment:
> In the original above code, in the foreach loop, `if (empty($value))` will call `continue;` if that happens on the last item in looping over `identifyingAttributes` then after the loop we have `empty($value)==1`.

Good catch! Thanks. I've fixed that by explicitly unsetting $value if it's empty. I also reversed the logic inside the foreach loop as you suggested. Not sure why I didn'tthink of that before; it's much cleaner now. Lots of nested if-loops to test all of the conditions, but I think it flows nicely, and I tried to comment it well. Hope this works better for you all and that you can include it in 2.4.

Reply:
I like the update. I did a quick think about how I might break it but if the attribute exists and is not empty it gets returned otherwise that var is `unset` again. So I don't see an unintended leak now.